mesonbuild/cargo/interpreter.py

```diff
@@ -578,7 +578,10 @@ def _enable_feature(self, pkg: PackageState, feature: str) -> None:
                         # This feature will be enabled only if that dependency
                         # is later added.
                         pkg.optional_deps_features[depname].add(dep_f)
-                else:
+                # Only enable features in dependencies if the dependency
+                # was added previously. The feature toggle does not imply a
+                # dependency exists, dev-dependencies are optional.
+                elif depname in pkg.manifest.dependencies:
```

Comment:
The dev-dependencies case is fixed with https://github.com/mesonbuild/meson/pull/13976. Here we want to enable the dependency, optional dep is handled above with the `?`.

Quoting the doc:

```
The "package-name/feature-name" syntax will also enable package-name if it is an optional dependency. Often this is not what you want. You can add a ? as in "package-name?/feature-name" which will only enable the given feature if something else enables the optional dependency.
```

Reply:
Are you sure the dev-dependencies case is fixed by that change?

what I’m saying is that there are crate features that toggle features on dev-dependencies. Dev dependencies should not be added to required dependencies for the package.

Reply:
I mean dev dependencies gets ignored because we don't support building examples/doc/etc yet.

Reply:
Dev dependencies are ignored, but dev dependency *feature toggles* are not. That's the whole point of this change: ignoring feature toggles for dev-dependencies and build-dependencies.

Reply:
> Are you sure the dev-dependencies case is fixed by that change?

Yes, dev dependencies are now ignored. We don't support yet building tests so we don't need them at all.
https://github.com/mesonbuild/meson/blob/master/mesonbuild/cargo/interpreter.py#L552

> what I’m saying is that there are crate features that toggle features on dev-dependencies.

Enabling feature on a dev dependency is also ignored there:
https://github.com/mesonbuild/meson/blob/master/mesonbuild/cargo/interpreter.py#L589

I think we are good now, but let me know if you think I missed something.

Reply:
My patch does basically the same thing as yours. Basically you have a `elif` and I have an extra `if` block. This PR was actually a bit more elegant, but it was missing other places that calls _add_dependency().